src/state-item.c

```diff
@@ -0,0 +1,417 @@
+//
+//  state-item.c
+//  bison
+//
+//  Created by Vincent Imbimbo on 10/15/19.
+//  Copyright © 2019 Vincent Imbimbo. All rights reserved.
+//
+
+#include <config.h>
+#include <stdlib.h>
+#include "state-item.h"
+#include "closure.h"
+#include "nullable.h"
+#include <time.h>
+#include <gl_linked_list.h>
+#include <gl_xlist.h>
+#include <assert.h>
+#include "getargs.h"
+
+size_t nstate_items;
+state_item_number *state_item_map;
+state_item *state_items;
+
+state_item_number *trans;
+bitsetv rev_trans;
+Hash_table *prods;
+Hash_table *rev_prods;
+
+
+// hash functions for index -> bitset hash maps
+typedef struct
+{
+    int key;
+    bitset l;
+} hash_pair;
+
+static size_t
+hash_pair_hasher(const void *sl, size_t max)
+{
+    int key = ((hash_pair*)sl)->key;
+    return key % max;
+}
+
+static bool
+hash_pair_comparator(const void *l, const void *r)
+{
+    return ((hash_pair *)l)->key == ((hash_pair *)r)->key;
+}
+
+static bitset
+hash_pair_lookup(Hash_table *tab, int key)
+{
+    hash_pair *l = xmalloc(sizeof(hash_pair));
+    l->key = key;
+    hash_pair *hp = (hash_pair *)hash_lookup(tab, l);
+    if (hp == NULL) return NULL;
+    return hp->l;
+}
+
+static void
+hash_pair_insert(Hash_table *tab, int key, bitset val)
+{
+    hash_pair *hp = xmalloc(sizeof(hash_pair));
+    hp->key = key;
+    hp->l = val;
+    if (!hash_insert(tab, hp))
+        xalloc_die();
+}
+
+/* return a state_item from a state's id and the offset of the item
+   within the state.
+ */
+state_item*
+state_item_lookup(state_number s, state_item_number off)
+{
+    return &state_items[state_item_index_lookup(s, off)];
+}
+
+static inline void
+state_item_set(state_item_number sidx, state *s, item_number off)
+{
+    state_item *si = state_items + sidx;
+    si->state = s;
+    si->item = &ritem[off];
+    si->reduction_offset = -1;
+    si->num = sidx;
+    trans[sidx] = -1;
+}
+/**
+ * Initialize state_items set
+ */
+static void
+init_state_items()
+{
+    nstate_items = 0;
+    bitsetv production_items = bitsetv_create(nstates, nritems, BITSET_SPARSE);
```

Comment:
Space before parens in every call.
You should definitely use clang-format to match the GNU style.
Each commit should be updated individually, so that we still have a nice series of commits, as you did.